src/core/qgsvectorfilewriter.cpp

```diff
@@ -3092,7 +3092,11 @@ QgsVectorFileWriter::WriterError QgsVectorFileWriter::prepareWriteAsVectorFormat
   {
     for ( int attrIdx : std::as_const( details.attributes ) )
     {
-      details.outputFields.append( details.sourceFields.at( attrIdx ) );
+      QgsField field = details.sourceFields.at( attrIdx );
+      if ( !options.attributesExportNames.isEmpty() &&
+           field.name() != options.attributesExportNames.at( attrIdx ) )
+        field.setName( options.attributesExportNames.at( attrIdx ) );
```

Comment:
```suggestion
      field.setName( options.attributesExportNames.value( attrIdx, field.name() ) );
```

This is a bit safer, in that it won't crash if the attributesExportName list size is smaller than the number of fields